helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java

```diff
@@ -197,6 +371,33 @@ public static ClusterModel generateClusterModelFromCurrentState(
     return toBeAssignedReplicas;
   }
 
+  // <partition, <state, instances set>>
+  private static Map<String, Map<String, Set<String>>> getValidStateInstanceMap(
+      ResourceAssignment assignment, Set<String> activeInstances) {
+    Map<String, Map<String, Set<String>>> stateInstanceMap = getStateInstanceMap(assignment);
+    // Filter to remove all invalid allocations that are not on the active instances.
+    stateInstanceMap.values().stream().forEach(stateMap -> stateMap.values().stream()
+        .forEach(instanceSet -> instanceSet.retainAll(activeInstances)));
+    return stateInstanceMap;
+  }
+
+  // <partition, <state, instances set>>
+  private static Map<String, Map<String, Set<String>>> getStateInstanceMap(
+      ResourceAssignment assignment) {
+    if (assignment == null) {
+      return Collections.emptyMap();
+    }
+    return assignment.getMappedPartitions().stream()
+        .collect(Collectors.toMap(partition -> partition.getPartitionName(), partition -> {
+          Map<String, Set<String>> stateInstanceMap = new HashMap<>();
+          assignment.getReplicaMap(partition).entrySet().stream().forEach(
+              stateMapEntry -> stateInstanceMap
+                  .computeIfAbsent(stateMapEntry.getValue(), key -> new HashSet<>())
+                  .add(stateMapEntry.getKey()));
+          return stateInstanceMap;
+        }));
```

Comment:
Are we iterating over the entire map twice here in the case of ValidStateInstanceMap calculation? 

Could we save some calculation by structuring this better? This just looks like functional programming.

Reply:
I tried to pass a nullable activeInstances so we only do the loop once. But that parameter does not look nice. Not a preferred coding style.
The performance concern here is minimum. The loops are doing once in each rebalance. So it is still O(1).